tgt_loop.cpp

```diff
@@ -256,6 +280,85 @@ static co_io_job __loop_handle_io_async(struct ublksrv_queue *q,
 	struct io_uring_cqe *cqe;
 	int ret;
 
+	const struct ublksrv_io_desc *iod = ublksrv_get_iod(q, tag);
+	unsigned ublk_op = ublksrv_get_op(iod);
+	unsigned long ioctl_op = 0;
+
+	switch (ublk_op) {
+	case UBLK_IO_OP_ZONE_OPEN:
+		ioctl_op = BLKOPENZONE;
+		break;
+	case UBLK_IO_OP_ZONE_CLOSE:
+		ioctl_op = BLKCLOSEZONE;
+		break;
+	case UBLK_IO_OP_ZONE_FINISH:
+		ioctl_op = BLKFINISHZONE;
+		break;
+	case UBLK_IO_OP_ZONE_RESET:
+		ioctl_op = BLKRESETZONE;
+		break;
+	}
+
+	if (ioctl_op) {
+		struct blk_zone_range range = {
+			.sector = iod->start_sector,
+			.nr_sectors = q->dev->tgt.zone_size_sectors,
+		};
+
+		ret = ioctl(q->dev->tgt.fds[1], ioctl_op, &range);
+
+    if (ret > 0) {
+      ret = 0;
+    }
+
+		ublksrv_complete_io(q, tag, ret);
+		co_return;
+	}
+
+  if (ublk_op == UBLK_IO_OP_DRV_IN) {
+		struct {
+			struct blk_zone_report report;
```

Comment:
This does not compile for me:

In file included from tgt_loop.cpp:5: /usr/include/linux/blkzoned.h: In function ‘co_io_job __loop_handle_io_async(ublksrv_queue*, ublk_io*, int)’: /usr/include/linux/blkzoned.h:133:25: error: flexible array member ‘blk_zone_report::zones’ not at end of ‘struct __loop_handle_io_async(ublksrv_queue*, ublk_io*, int)::<unnamed>’ 133 |         struct blk_zone zones[]; |                         ^~~~~ tgt_loop.cpp:321:41: note: next member ‘blk_zone __loop_handle_io_async(ublksrv_queue*, ublk_io*, int)::<unnamed struct>::zone’ declared here 321 |                         struct blk_zone zone; |                                         ^~~~ tgt_loop.cpp:319:24: note: in the definition of ‘struct __loop_handle_io_async(ublksrv_queue*, ublk_io*, int)::<unnamed>’ 319 |                 struct { |                        ^ In file included from tgt_loop.cpp:5: /usr/include/linux/blkzoned.h:133:25: error: flexible array member ‘blk_zone_report::zones’ not at end of ‘struct __loop_handle_io_async(ublksrv_queue*, ublk_io*, int)::_ZL22__loop_handle_io_asyncP13ublksrv_queueP7ublk_ioi.frame’ 133 |         struct blk_zone zones[]; |                         ^~~~~ tgt_loop.cpp:321:41: note: next member ‘blk_zone __loop_handle_io_async(ublksrv_queue*, ublk_io*, int)::<unnamed struct>::zone’ declared here 321 |                         struct blk_zone zone; |                                         ^~~~ tgt_loop.cpp:384:1: note: in the definition of ‘struct __loop_handle_io_async(ublksrv_queue*, ublk_io*, int)::_ZL22__loop_handle_io_asyncP13ublksrv_queueP7ublk_ioi.frame’ 384 | } | ^ 

Why not just do this properly: allocate space for the requested number of zones, then get all those reported through the ioctl, just like in https://github.com/westerndigitalcorporation/libzbd/blob/master/lib/zbd.c#L520

Reply:
Getting all the zones (or up to a max page worth maybe) is a good idea. It is however not related to the reason the code is not building for you. Are you using an old compiler?

Reply:
@yhr I pushed a fix for that warning, can you check if it resolves your issue?

Reply:
The latest push looks better and compiles, thanks!

I think we have discussed this before, but what is the reason for not generating a full report? For a device with a thousand zones the kernel-userspace ping-pong game will be a long one. Packing up a report in a page will cut down the report calls with a factor of 17 , but it would be good to avoid any unneeded roundtrips.

Reply:
> I think we have discussed this before, but what is the reason for not generating a full report? For a device with a thousand zones the kernel-userspace ping-pong game will be a long one. Packing up a report in a page will cut down the report calls with a factor of 17 , but it would be good to avoid any unneeded roundtrips.

I agree there is an opportunity for optimizing the zone report. But to avoid premature optimization, I did not include this in the the current version of the patch. I think the round trip to user space and back is around 1us, so we would waste 1ms for a device with 1k zones. However, I think this is completely dwarfed by the time current devices spend on producing a full zone report anyway? Also this would probably only be invoked during initialization phases, so it does not really matter. We can always change it?

We should encode the number of zones, one or more, in the kernel<->userspace api now though. That might be difficult to change later if we want to.

Reply:
I think it's important to make sure that the uabi is reasonably future-proof, so reporting more than one zone at a time must be there I believe. As the code is today, the per-zone IOCTL will generate a l lot of overhead for real devices. I'll look over the kernel patch next -  the uabi is probably best discussed on the block list .